File: 10_flutter_and_native_samples/ios/Runner/BiometricAuthChannel.swift
import Flutter
import LocalAuthentication
import UIKit

/// Bridges Flutter MethodChannel calls to LocalAuthentication.
///
/// iOS evaluates biometrics and returns only the authentication result.
final class BiometricAuthChannel {
  static let channelName = "examples.native_biometric_auth/biometrics"

  private let channel: FlutterMethodChannel
  private var authenticationInProgress = false

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
  }

  func register() {
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
  }

  func unregister() {
    channel.setMethodCallHandler(nil)
    authenticationInProgress = false
  }

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getBiometricStatus":
      result(getBiometricStatus())
    case "authenticate":
      authenticate(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func getBiometricStatus() -> [String: Any?] {
    let context = LAContext()
    var error: NSError?

    // Probe availability without presenting a system prompt.
    let canEvaluate = context.canEvaluatePolicy(
      .deviceOwnerAuthenticationWithBiometrics,
      error: &error
    )

    if canEvaluate {
      return [
        "available": true,
        "type": biometricTypeName(for: context.biometryType),
        "reason": nil,
      ]
    }

    let reason = mapAvailabilityError(error)
    return [
      "available": false,
      "type": "none",
      "reason": reason,
    ]
  }

  private func authenticate(call: FlutterMethodCall, result: @escaping FlutterResult) {
    if authenticationInProgress {
      result(
        authFailure(
          errorCode: "alreadyInProgress",
          errorMessage: "A biometric authentication request is already in progress."
        )
      )
      return
    }

    guard
      let args = call.arguments as? [String: Any],
      let reason = args["reason"] as? String,
      !reason.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    else {
      result(
        authFailure(
          errorCode: "invalidArguments",
          errorMessage: "authenticate requires a non-empty reason string."
        )
      )
      return
    }

    let status = getBiometricStatus()
    let available = status["available"] as? Bool ?? false
    if !available {
      let unavailableReason = status["reason"] as? String ?? "unavailable"
      result(
        authFailure(
          errorCode: unavailableReason,
          errorMessage: "Biometric authentication is not available (\(unavailableReason))."
        )
      )
      return
    }

    authenticationInProgress = true

    // Create a fresh LAContext for each authentication attempt.
    let context = LAContext()
    context.localizedCancelTitle = "Cancel"

    // iOS evaluates biometrics and returns only the authentication result.
    context.evaluatePolicy(
      .deviceOwnerAuthenticationWithBiometrics,
      localizedReason: reason
    ) { [weak self] success, error in
      DispatchQueue.main.async {
        guard let self else {
          return
        }

        self.authenticationInProgress = false

        if success {
          result([
            "success": true,
            "errorCode": NSNull(),
            "errorMessage": NSNull(),
          ])
          return
        }

        let mapped = self.mapAuthenticationError(error)
        result(mapped)
      }
    }
  }

  private func biometricTypeName(for type: LABiometryType) -> String {
    switch type {
    case .touchID:
      return "fingerprint"
    case .faceID:
      return "face"
    case .opticID:
      return "iris"
    case .none:
      return "none"
    @unknown default:
      return "unknown"
    }
  }

  private func mapAvailabilityError(_ error: NSError?) -> String {
    guard let error else {
      return "unknown"
    }

    switch error.code {
    case LAError.biometryNotAvailable.rawValue:
      return "notSupported"
    case LAError.biometryNotEnrolled.rawValue:
      return "notEnrolled"
    case LAError.biometryLockout.rawValue:
      return "temporarilyLocked"
    case LAError.passcodeNotSet.rawValue:
      return "notSupported"
    default:
      return "unavailable"
    }
  }

  private func mapAuthenticationError(_ error: Error?) -> [String: Any?] {
    guard let laError = error as? LAError else {
      return authFailure(
        errorCode: "unknown",
        errorMessage: error?.localizedDescription ?? "Biometric authentication failed."
      )
    }

    switch laError.code {
    case .userCancel:
      return authFailure(
        errorCode: "userCanceled",
        errorMessage: "Authentication was canceled by the user"
      )
    case .appCancel, .systemCancel:
      return authFailure(
        errorCode: "systemCanceled",
        errorMessage: "Authentication was canceled by the system"
      )
    case .authenticationFailed:
      return authFailure(
        errorCode: "authenticationFailed",
        errorMessage: "Biometric authentication failed."
      )
    case .biometryNotAvailable:
      return authFailure(
        errorCode: "notSupported",
        errorMessage: "Biometric authentication is not supported on this device."
      )
    case .biometryNotEnrolled:
      return authFailure(
        errorCode: "notEnrolled",
        errorMessage: "No biometrics are enrolled on this device."
      )
    case .biometryLockout:
      return authFailure(
        errorCode: "temporarilyLocked",
        errorMessage: "Biometric authentication is temporarily locked."
      )
    case .passcodeNotSet:
      return authFailure(
        errorCode: "notSupported",
        errorMessage: "Device passcode is not set; biometrics cannot be used."
      )
    case .userFallback:
      // Policy excludes device credentials; treat fallback as cancel.
      return authFailure(
        errorCode: "userCanceled",
        errorMessage: "Authentication was canceled by the user"
      )
    case .invalidContext, .notInteractive:
      return authFailure(
        errorCode: "unavailable",
        errorMessage: laError.localizedDescription
      )
    default:
      return authFailure(
        errorCode: "unknown",
        errorMessage: laError.localizedDescription
      )
    }
  }

  private func authFailure(errorCode: String, errorMessage: String) -> [String: Any?] {
    [
      "success": false,
      "errorCode": errorCode,
      "errorMessage": errorMessage,
    ]
  }
}

File: 10_flutter_and_native_samples/android/app/src/main/kotlin/com/example/flutter_and_native_samples/BiometricAuthChannel.kt
package com.example.flutter_and_native_samples

import android.content.pm.PackageManager
import android.os.Build
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Bridges Flutter MethodChannel calls to AndroidX BiometricPrompt.
 *
 * Android owns and displays the biometric system prompt.
 */
class BiometricAuthChannel(
    private val activity: FragmentActivity,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {
    private val channel =
        MethodChannel(messenger, "examples.native_biometric_auth/biometrics")

    private var authenticationInProgress = false
    private var pendingResult: MethodChannel.Result? = null
    private var activePrompt: BiometricPrompt? = null

    fun register() {
        channel.setMethodCallHandler(this)
    }

    fun unregister() {
        channel.setMethodCallHandler(null)
        activePrompt?.cancelAuthentication()
        complete(
            authFailure(
                errorCode = "systemCanceled",
                errorMessage = "Authentication was canceled by the system",
            ),
        )
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getBiometricStatus" -> result.success(getBiometricStatus())
            "authenticate" -> authenticate(call, result)
            else -> result.notImplemented()
        }
    }

    private fun getBiometricStatus(): Map<String, Any?> {
        val biometricManager = BiometricManager.from(activity)
        val authenticity =
            biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG)

        return when (authenticity) {
            BiometricManager.BIOMETRIC_SUCCESS ->
                mapOf(
                    "available" to true,
                    "type" to detectBiometricType(),
                    "reason" to null,
                )

            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE,
            BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED,
            BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED,
            ->
                unavailableStatus(reason = "notSupported")

            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED ->
                unavailableStatus(reason = "notEnrolled")

            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE ->
                unavailableStatus(reason = "unavailable")

            BiometricManager.BIOMETRIC_STATUS_UNKNOWN ->
                unavailableStatus(reason = "unknown")

            else -> unavailableStatus(reason = "unknown")
        }
    }

    private fun authenticate(call: MethodCall, result: MethodChannel.Result) {
        if (authenticationInProgress) {
            result.success(
                authFailure(
                    errorCode = "alreadyInProgress",
                    errorMessage = "A biometric authentication request is already in progress.",
                ),
            )
            return
        }

        val reason = call.argument<String>("reason")
        if (reason.isNullOrBlank()) {
            result.success(
                authFailure(
                    errorCode = "invalidArguments",
                    errorMessage = "authenticate requires a non-empty reason string.",
                ),
            )
            return
        }

        val status = getBiometricStatus()
        val available = status["available"] as? Boolean ?: false
        if (!available) {
            val unavailableReason = status["reason"] as? String ?: "unavailable"
            result.success(
                authFailure(
                    errorCode = unavailableReason,
                    errorMessage = "Biometric authentication is not available ($unavailableReason).",
                ),
            )
            return
        }

        authenticationInProgress = true
        pendingResult = result

        val executor = ContextCompat.getMainExecutor(activity)
        val prompt =
            BiometricPrompt(
                activity,
                executor,
                object : BiometricPrompt.AuthenticationCallback() {
                    override fun onAuthenticationSucceeded(
                        result: BiometricPrompt.AuthenticationResult,
                    ) {
                        complete(
                            mapOf(
                                "success" to true,
                                "errorCode" to null,
                                "errorMessage" to null,
                            ),
                        )
                    }

                    override fun onAuthenticationError(
                        errorCode: Int,
                        errString: CharSequence,
                    ) {
                        val mapped = mapPromptError(errorCode, errString.toString())
                        complete(mapped)
                    }

                    override fun onAuthenticationFailed() {
                        // Intermediate failure; the system prompt stays open.
                    }
                },
            )

        activePrompt = prompt

        val promptInfo =
            BiometricPrompt.PromptInfo.Builder()
                .setTitle("Biometric authentication")
                .setSubtitle(reason)
                .setNegativeButtonText("Cancel")
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                .build()

        // Android owns and displays the biometric system prompt.
        prompt.authenticate(promptInfo)
    }

    private fun mapPromptError(
        errorCode: Int,
        errString: String,
    ): Map<String, Any?> {
        val mappedCode =
            when (errorCode) {
                BiometricPrompt.ERROR_USER_CANCELED,
                BiometricPrompt.ERROR_NEGATIVE_BUTTON,
                -> "userCanceled"

                BiometricPrompt.ERROR_CANCELED -> "systemCanceled"

                BiometricPrompt.ERROR_LOCKOUT -> "temporarilyLocked"

                BiometricPrompt.ERROR_LOCKOUT_PERMANENT -> "permanentlyLocked"

                BiometricPrompt.ERROR_NO_BIOMETRICS -> "notEnrolled"

                BiometricPrompt.ERROR_HW_NOT_PRESENT,
                BiometricPrompt.ERROR_NO_DEVICE_CREDENTIAL,
                -> "notSupported"

                BiometricPrompt.ERROR_HW_UNAVAILABLE,
                BiometricPrompt.ERROR_VENDOR,
                BiometricPrompt.ERROR_UNABLE_TO_PROCESS,
                BiometricPrompt.ERROR_TIMEOUT,
                -> "unavailable"

                BiometricPrompt.ERROR_NO_SPACE -> "unavailable"

                else -> "authenticationFailed"
            }

        val message =
            when (mappedCode) {
                "userCanceled" -> "Authentication was canceled by the user"
                "systemCanceled" -> "Authentication was canceled by the system"
                else -> errString.ifBlank { "Biometric authentication failed ($mappedCode)." }
            }

        return authFailure(errorCode = mappedCode, errorMessage = message)
    }

    private fun detectBiometricType(): String {
        val packageManager = activity.packageManager
        val fingerprint = packageManager.hasSystemFeature(PackageManager.FEATURE_FINGERPRINT)
        val face =
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q &&
                packageManager.hasSystemFeature(PackageManager.FEATURE_FACE)
        val iris =
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q &&
                packageManager.hasSystemFeature(PackageManager.FEATURE_IRIS)

        val enabledCount = listOf(fingerprint, face, iris).count { it }
        return when {
            enabledCount > 1 -> "multiple"
            fingerprint -> "fingerprint"
            face -> "face"
            iris -> "iris"
            else -> "unknown"
        }
    }

    private fun unavailableStatus(reason: String): Map<String, Any?> =
        mapOf(
            "available" to false,
            "type" to "none",
            "reason" to reason,
        )

    private fun authFailure(
        errorCode: String,
        errorMessage: String,
    ): Map<String, Any?> =
        mapOf(
            "success" to false,
            "errorCode" to errorCode,
            "errorMessage" to errorMessage,
        )

    private fun complete(payload: Map<String, Any?>) {
        val result = pendingResult ?: return
        clearInProgressState()
        result.success(payload)
    }

    private fun clearInProgressState() {
        authenticationInProgress = false
        pendingResult = null
        activePrompt = null
    }
}
